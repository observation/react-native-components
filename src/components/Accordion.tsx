import React, { useState } from 'react'
import { View } from 'react-native'

// @ts-ignore
import { Collapse, CollapseHeader, CollapseBody } from 'accordion-collapse-react-native'

import Log from '../lib/Log'
import { unsafeLayoutAnimation } from '../lib/Utils'

type Item = {
  title: string
}

type Props<T extends Item> = {
  list: T[]
  header: (item: T, index: number, isActive: boolean) => void
  footer: (item: T, index: number, isActive: boolean) => void
  body: (item: T) => JSX.Element
  onOpen: (index: number) => void
  ListEmptyComponent?: () => JSX.Element
}

const Accordion = <T extends Item>({ list, header, footer, body, onOpen, ListEmptyComponent }: Props<T>) => {
  const [activeIndex, setActiveIndex] = useState<number | null>(null)

  const onToggle = (index: number, isExpanded: boolean) => {
    Log.debug('Accordion:onToggle', index, isExpanded)
    unsafeLayoutAnimation('Accordion:onToggle')
    if (isExpanded) {
      setActiveIndex(index)
      onOpen(index)
    } else if (activeIndex === index) {
      setActiveIndex(null)
    }
  }

  return (
    <View>
      {list.length === 0
        ? ListEmptyComponent?.()
        : list.map((item, index: number) => (
            <Collapse
              key={item.title}
              isExpanded={index === activeIndex}
              onToggle={(isExpanded: boolean) => onToggle(index, isExpanded)}
            >
              <CollapseHeader>
                <>
                  {header(item, index, index === activeIndex)}
                  {footer(item, index, index === activeIndex)}
                </>
              </CollapseHeader>
              <CollapseBody>{body(item)}</CollapseBody>
            </Collapse>
          ))}
    </View>
  )
}

export default Accordion
