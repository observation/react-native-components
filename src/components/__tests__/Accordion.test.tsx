import React from 'react'
import { Text } from 'react-native'

import { fireEvent, render } from '@testing-library/react-native'

import Accordion from '../Accordion'

type AccordionItem = {
  title: string
}

const header = (item: AccordionItem) => <Text>{`Header ${item.title}`}</Text>
const footer = (item: AccordionItem) => <Text>{`Footer ${item.title}`}</Text>
const body = (item: AccordionItem) => <Text>{`Body ${item.title}`}</Text>
const onOpen = jest.fn()

const list: AccordionItem[] = [{ title: '1st item' }, { title: '2nd item' }, { title: '3rd item' }]

describe('Accordion', () => {
  beforeEach(() => {
    jest.clearAllMocks()
  })

  describe('Rendering', () => {
    test('Normal render', () => {
      const { toJSON } = render(
        <Accordion<AccordionItem> list={list} header={header} footer={footer} body={body} onOpen={onOpen} />,
      )
      expect(toJSON()).toMatchSnapshot()
    })

    test('Empty list', () => {
      const { toJSON } = render(
        <Accordion<AccordionItem> list={[]} header={header} footer={footer} body={body} onOpen={onOpen} />,
      )
      expect(toJSON()).toMatchSnapshot()
    })
  })

  describe('Interaction', () => {
    test('Clicking on the 1st item, shows the 1st body', async () => {
      // GIVEN
      const { getByText, queryByText, toJSON } = render(
        <Accordion<AccordionItem> list={list} header={header} footer={footer} body={body} onOpen={onOpen} />,
      )

      // WHEN
      fireEvent.press(getByText('Header 1st item'))

      // THEN
      expect(onOpen).toBeCalledTimes(1)
      expect(queryByText('Body 1st item')).toBeTruthy()
      expect(toJSON()).toMatchSnapshot()
    })

    test('Clicking on an 1st item, then the 2nd item, shows the 2nd body', async () => {
      // GIVE
      const { getByText, queryByText, toJSON } = render(
        <Accordion<AccordionItem> list={list} header={header} footer={footer} body={body} onOpen={onOpen} />,
      )

      // WHEN
      fireEvent.press(getByText('Header 1st item'))
      fireEvent.press(getByText('Header 2nd item'))

      // THEN
      expect(onOpen).toBeCalledTimes(2)
      expect(queryByText('Body 2nd item')).toBeTruthy()

      expect(toJSON()).toMatchSnapshot()
    })

    test('Clicking on an 1st item twice, shows no body', async () => {
      // GIVE
      const { queryByText, getByText, toJSON } = render(
        <Accordion<AccordionItem> list={list} header={header} footer={footer} body={body} onOpen={onOpen} />,
      )

      // WHEN
      fireEvent.press(getByText('Header 1st item'))
      fireEvent.press(getByText('Header 1st item'))

      // THEN
      expect(onOpen).toBeCalledTimes(1)
      expect(queryByText('Body 1st item')).not.toBeTruthy()
      expect(toJSON()).toMatchSnapshot()
    })
  })
})
