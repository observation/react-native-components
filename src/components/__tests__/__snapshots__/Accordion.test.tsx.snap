// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordion Interaction Clicking on an 1st item twice, shows no body 1`] = `
<View>
  <View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": false,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
    >
      <View>
        <Text>
          Header 1st item
        </Text>
        <Text>
          Footer 1st item
        </Text>
      </View>
    </View>
  </View>
  <View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": false,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
    >
      <View>
        <Text>
          Header 2nd item
        </Text>
        <Text>
          Footer 2nd item
        </Text>
      </View>
    </View>
  </View>
  <View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": false,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
    >
      <View>
        <Text>
          Header 3rd item
        </Text>
        <Text>
          Footer 3rd item
        </Text>
      </View>
    </View>
  </View>
</View>
`;

exports[`Accordion Interaction Clicking on an 1st item, then the 2nd item, shows the 2nd body 1`] = `
<View>
  <View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": false,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
    >
      <View>
        <Text>
          Header 1st item
        </Text>
        <Text>
          Footer 1st item
        </Text>
      </View>
    </View>
  </View>
  <View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": false,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
    >
      <View>
        <Text>
          Header 2nd item
        </Text>
        <Text>
          Footer 2nd item
        </Text>
      </View>
    </View>
    <View>
      <Text>
        Body 2nd item
      </Text>
    </View>
  </View>
  <View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": false,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
    >
      <View>
        <Text>
          Header 3rd item
        </Text>
        <Text>
          Footer 3rd item
        </Text>
      </View>
    </View>
  </View>
</View>
`;

exports[`Accordion Interaction Clicking on the 1st item, shows the 1st body 1`] = `
<View>
  <View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": false,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
    >
      <View>
        <Text>
          Header 1st item
        </Text>
        <Text>
          Footer 1st item
        </Text>
      </View>
    </View>
    <View>
      <Text>
        Body 1st item
      </Text>
    </View>
  </View>
  <View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": false,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
    >
      <View>
        <Text>
          Header 2nd item
        </Text>
        <Text>
          Footer 2nd item
        </Text>
      </View>
    </View>
  </View>
  <View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": false,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
    >
      <View>
        <Text>
          Header 3rd item
        </Text>
        <Text>
          Footer 3rd item
        </Text>
      </View>
    </View>
  </View>
</View>
`;

exports[`Accordion Rendering Empty list 1`] = `<View />`;

exports[`Accordion Rendering Normal render 1`] = `
<View>
  <View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": false,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
    >
      <View>
        <Text>
          Header 1st item
        </Text>
        <Text>
          Footer 1st item
        </Text>
      </View>
    </View>
  </View>
  <View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": false,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
    >
      <View>
        <Text>
          Header 2nd item
        </Text>
        <Text>
          Footer 2nd item
        </Text>
      </View>
    </View>
  </View>
  <View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": false,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
    >
      <View>
        <Text>
          Header 3rd item
        </Text>
        <Text>
          Footer 3rd item
        </Text>
      </View>
    </View>
  </View>
</View>
`;
